sampotts/plyr
# Improve captions handling for streaming

Relates to #971

* Adds option to watch caption track changes and update language options
* Ensure languagechange and captionsenabled/captionsdisabled events are sent on initial setup, and that captionsenabled/captionsdisabled is only triggered on change.
* Ensure language is set in case the track is added after initialization
* Optimizations for code reuse.

`instance.toggleCaptions` will now ensure the elements are updated with the right classes/state even if they already were before. The reason for this is code reuse. I was able to remove the`captions.show` method this way. Without the changes to `instance.toggleCaptions` however it would not work if you load a new source in the same instance.

Another thing that may need explaining is `this.language` is being set in the new `captions.update` method (when it hasn't been set already). This is needed because setting `this.language` will rightfully bail if the language doesn't have a track (and log a warning which may need to be upgraded now). The track for the language may be added after initializing Plyr, so it has to try on every update until it has a match.

How to test:
```js
new Plyr(video, {captions: {update: true}});
video.addTextTrack('subtitles', 'Volapük', 'vo');
```

Hls.js introduces another dimension of issues to the table, and it's not really ours to fix, but if you want to test it, here's a new simplified jsbin (using this branch) with Plyr + hls.js and two kinds of captions:
http://jsbin.com/mayikihosu/edit?js,output

Not in this PR:
* It doesn't fix the css issue (I think?) causing #980, however it does ensure `player.language` isn't something that doesn't have a track. So the last bullet would be `undefined` not `es`
* Keeping an internal reference of the current track index - This would be nice for other streaming libraries as well since it seems to be how they operate (the same applies to quality options). The method I'm running in the `languagechange` listener in the jsfiddle examples is not safe without this, if there are duplicate tracks per language (like hard-of-hearing subtitles).
